onnxruntime/python/onnxruntime_pybind_mlvalue.cc

```diff
@@ -196,13 +212,95 @@
 
   auto hit = id_to_allocator_map->find(id);
   if (hit == id_to_allocator_map->end()) {
-    auto dml_allocator = std::make_shared<Dml::DmlExternalBufferAllocator>(id);
+    constexpr uint32_t device_id = 0;
+    auto d3d12_device = onnxruntime::DMLProviderFactoryCreator::CreateD3D12Device(device_id, false);
+    auto dml_device = onnxruntime::DMLProviderFactoryCreator::CreateDMLDevice(d3d12_device.Get());
+
+    D3D12_COMMAND_QUEUE_DESC cmd_queue_desc = {};
+    cmd_queue_desc.Type = D3D12_COMMAND_LIST_TYPE_DIRECT;
+    cmd_queue_desc.Flags = D3D12_COMMAND_QUEUE_FLAG_DISABLE_GPU_TIMEOUT;
+
+    ComPtr<ID3D12CommandQueue> cmd_queue;
+    ORT_THROW_IF_FAILED(d3d12_device->CreateCommandQueue(&cmd_queue_desc, IID_PPV_ARGS(cmd_queue.ReleaseAndGetAddressOf())));
+
+    auto context = std::make_shared<Dml::ExecutionContext>(d3d12_device.Get(), dml_device.Get(), cmd_queue.Get());
+
+    // We leak the upload and readback heaps to keep them alive, just like the map
```

Comment:
>just like the map [](http://example.com/codeflow?start=65&length=17)

😿 How often is `GetDmlAllocator` called? I hate to think that every time I rerun my Stable Diffusion model that it leaks more expensive GPU heaps 😬.

Reply:
The number of times it gets called doesn't matter. The heaps are created and cached once per process, so subsequent calls will get cached heaps.

Reply:
True, it's a [magic `static`](https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables).